controllers/stardoginstance_controller.go

```diff
@@ -7,36 +7,293 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"k8s.io/apimachinery/pkg/types"
+	"net/url"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"time"
 
 	"github.com/go-logr/logr"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	stardogv1alpha1 "github.com/vshn/stardog-userrole-operator/api/v1alpha1"
+	. "github.com/vshn/stardog-userrole-operator/api/v1alpha1"
+	rest "github.com/vshn/stardog-userrole-operator/stardogrest"
 )
 
 // StardogInstanceReconciler reconciles a StardogInstance object
-type StardogInstanceReconciler struct {
-	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
-}
+type (
+	StardogInstanceReconciler struct {
+		client.Client
+		Log    logr.Logger
+		Scheme *runtime.Scheme
+		ReconcileInterval time.Duration
+	}
+)
+
+const stardogUsersFinalizer = "finalizer.stardog.users"
+const stardogRolesFinalizer = "finalizer.stardog.roles"
 
 // +kubebuilder:rbac:groups=stardog.vshn.ch,resources=stardoginstances,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=stardog.vshn.ch,resources=stardoginstances/status,verbs=get;update;patch
-
 func (r *StardogInstanceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("stardoginstance", req.NamespacedName)
+	namespace := req.NamespacedName
+	_ = r.Log.WithValues("stardoginstance", namespace)
+
+    ctx := context.Background()
+    stardogInstance := &StardogInstance{}
+
+    err := r.Client.Get(ctx, namespace, stardogInstance)
+    if err != nil {
+        if apierrors.IsNotFound(err) {
+            r.Log.Info("StardogInstance not found, ignoring reconcile.", "StardogInstance", namespace)
+            return ctrl.Result{Requeue: false}, nil
+        }
+        r.Log.Error(err, "Could not retrieve StardogInstance.", "StardogInstance", namespace)
+        return ctrl.Result{Requeue: true, RequeueAfter: r.ReconcileInterval}, err
+    }
+
+	rc := &ReconciliationContext{
+		ctx:        ctx,
+		cfg:        stardogInstance,
+		conditions: make(map[StardogConditionType]StardogCondition),
+	}
+
+	r.Log.Info("reconciling", getLoggingKeysAndValuesForStardogInstance(stardogInstance)...)
+
+	//TODO find out how to delete when credentials do not work or the StardogInstance never connected to a Stardog DB
```

Comment:
Need some input here

Reply:
I would make it simple: Add the "Terminating" and "Errored" conditions and leave the finalizers until

1. the admin manually fixes the problem or 
2. removes the finalizer if the admin is certain that the object can be deleted

Because we simply cannot know whether they have been working previously or not. The credentials could also have been manually changed in Stardog without changing the secret, in which we land in the same situation.

Reply:
1) The admin may never be able to fix the problem - the current implementation works like that.
2) how do I figure out that the admin is certain the object can be deleted, there's no force argument as we talked last time

We need to find a loophole or something that would allow to delete a spec right away

Reply:
There are things which we just can't (and shouldn't) control. Like:
1. User uses the stardog-cli inside the pods to change admin password -> Operator does not know this and will fail all next reconciliations
2. The User deletes the stardog instance helm release without deleting the Instance CRD object before. Obviously deleting instance thanks to the finalizer will be forever blocked since the operator tries to connect to it.
3. The user changes the password in the secret -> Operator cannot change the admin password on stardog anymore, since it can't authorize anymore.

In essence: There are tons of different scenarios that can go wrong. so don't overthink it and just expect errors, we don't need to solve them for the user.

Removing the finalizer manually is very easy: `kubectl edit stardoginstance <name>` and then remove the `finalizers` array in the YAML, save & K8s deletes it. We don't need a "loophole" (whatever that is) to delete the spec right away, the tools are there already :smile:

Reply:
Ah ok, didn't know that we can manually delete the finalizers, than it's fine